File: project/jimmer-sql-test-model/src/main/java/org/babyfish/jimmer/sql/model/pg/TagsScalarProvider.java
package org.babyfish.jimmer.sql.model.pg;

import org.babyfish.jimmer.sql.runtime.ScalarProvider;
import org.jspecify.annotations.NonNull;

import java.util.List;

import static org.babyfish.jimmer.jackson.codec.JsonCodec.jsonCodec;

public class TagsScalarProvider implements ScalarProvider<List<String>, String> {
    @Override
    public List<String> toScalar(@NonNull String sqlValue) throws Exception {
        return jsonCodec().readerForListOf(String.class).read(sqlValue);
    }

    @Override
    public String toSql(@NonNull List<String> scalarValue) throws Exception {
        return jsonCodec().writer().writeAsString(scalarValue);
    }

    @Override
    public boolean isJsonScalar() {
        return true;
    }
}

File: project/jimmer-sql-test/jimmer-sql-test-model-kotlin/src/main/kotlin/org/babyfish/jimmer/sql/kt/model/provider/TagsScalarProvider.kt
package org.babyfish.jimmer.sql.kt.model.provider

import org.babyfish.jimmer.sql.runtime.ScalarProvider
import java.util.EnumSet

class TagsScalarProvider : ScalarProvider<EnumSet<Tag>, Int> {

    override fun toScalar(sqlValue: Int): EnumSet<Tag> {
        val tags = EnumSet.noneOf(Tag::class.java)
        for (tag in enumValues<Tag>()) {
            if (sqlValue and (1 shl tag.ordinal) != 0) {
                tags.add(tag)
            }
        }
        return tags
    }

    override fun toSql(scalarValue: EnumSet<Tag>): Int {
        var bits = 0
        for (tag in scalarValue) {
            bits = bits or (1 shl tag.ordinal)
        }
        return bits
    }
}